### DISCUSSION TITLE
`query_expression` in `WHERE`, `ORDER BY` sections

### DISCUSSION THREAD

**Discussion by swimmwatch** [sqlalchemy/sqlalchemy#10558] [Usage Questions] [answered]
Hi! I read about `query_expression` from [documentation](https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#orm-queryguide-with-expression). And it says:

> The mapped attribute cannot be applied to other parts of the query, such as the WHERE clause, the ORDER BY clause

I would like to know why this is so. I think it would be very useful feature.

**Comment by zzzeek** [comment, marked as answer]
`with_expression` is a loader option; it's purpose is to indicate how the ORM should populate ORM objects that will be returned from `session.execute`.   loader options are only processed for actual ORM objects that are to be returned.

that is, taking a different loader option, `defer`, this `defer` option does nothing:


    stmt = select(User.id, User.name).options(defer(User.fullname))

it does nothing because the statement is requesting `Row` objects that contain `id` and `name`, and will not contain `User` objects.   it's in fact mostly ignored (if not raises an error at this point).

by contrast, the option is now significant here:

    stmt = select(User).options(defer(User.fullname))

since the statement has a full `User` object, that implies a SELECT of all the columns in the user table and that they should be applied to the attributes on each `User` object returned.   That's where loader options take place, and nowhere else.


The presence of `query_expression` arose from the fact that people wanted to get back `User` objects and be able to dynamically alter an expression that is applied to the object, this is the purview of loader options, so that's where it takes place.

but just like `defer()`, if you made a query like this:

    stmt = select(User.id, User.name).options(with_expression(User.fullname, "some fullname"))

`with_expression` is irrelevant to that query just like `defer(User.fullname)` is.      Hence loader options are not the appropriate place to affect the WHERE / ORDER BY clauses of statements.  loader options don't affect "the SQL" per se, they indicate a behavior to take place as full objects are loaded only.

**Reply by swimmwatch** [reply]
Thanks for detailed answer!
